ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala

```diff
@@ -3,52 +3,340 @@
+        } yield {
+          val agreement = assertSingleton(
+            "SemanticPaintCounterOffer",
+            createdEvents(tree).filter(_.getTemplateId == Tag.unwrap(PaintAgree.id)))
+          assertEquals(
+            "Paint agreement parameters",
+            agreement.getCreateArguments,
+            Record(
+              recordId = Some(Tag.unwrap(PaintAgree.id)),
+              fields = Seq(
+                RecordField("painter", Some(Value(Value.Sum.Party(Tag.unwrap(painter))))),
+                RecordField("houseOwner", Some(Value(Value.Sum.Party(Tag.unwrap(houseOwner)))))
+              )
+            )
+          )
+        }
+    }
+
+  private[this] val partialSignatories =
+    LedgerTest(
+      "SemanticPartialSignatories",
+      "A signatory should not be able to create a contract on behalf of two parties") { ledger =>
+      for {
+        Vector(houseOwner, painter) <- ledger.allocateParties(2)
+        failure <- ledger.create(houseOwner, PaintAgree(painter, houseOwner)).failed
+      } yield {
+        assertGrpcError(failure, Status.Code.INVALID_ARGUMENT, "requires authorizers")
+      }
+    }
+
+  private[this] val acceptOnBehalf =
+    LedgerTest(
+      "SemanticAcceptOnBehalf",
+      "It should not be possible to exercise a choice without the consent of the controller") {
+      ledger =>
+        for {
+          Vector(bank, houseOwner, painter) <- ledger.allocateParties(3)
+          iou <- ledger.create(bank, Iou(bank, houseOwner, onePound))
+          offer <- ledger.create(painter, PaintOffer(painter, houseOwner, bank, onePound))
+          failure <- ledger.exercise(painter, offer.exercisePaintOffer_Accept(_, iou)).failed
+        } yield {
+          assertGrpcError(failure, Status.Code.INVALID_ARGUMENT, "couldn't find contract")
+        }
+    }
+
+  /*
+   * Privacy
+   *
+   * Visibility of contracts we fetch. Since the Ledger API has
+   * no fetch operation built-in, we use a template with a choice
+   * that causes the fetch.
+   */
+
+  private[this] val privacyProjections =
+    LedgerTest(
+      "SemanticPrivacyProjections",
+      "Test visibility via contract fetches for the paint-offer flow") { ledger =>
+      for {
+        Vector(bank, houseOwner, painter) <- ledger.allocateParties(3)
+        iou <- ledger.create(bank, Iou(bank, houseOwner, onePound))
+
+        // The IOU should be visible only to the payer and the owner
+        _ <- fetchIou(ledger, bank, iou)
+        _ <- fetchIou(ledger, houseOwner, iou)
+        _ <- fetchIou(ledger, painter, iou).failed
+
+        offer <- ledger.create(painter, PaintOffer(painter, houseOwner, bank, onePound))
+
+        // The house owner and the painter can see the offer but the bank can't
+        _ <- fetchPaintOffer(ledger, houseOwner, offer)
+        _ <- fetchPaintOffer(ledger, painter, offer)
+        _ <- fetchPaintOffer(ledger, bank, offer).failed
+
+        // Note that the new IOU was divulged to the house owner
```

Comment:
@stefanobaghino-da : it does not seem to me that we are exploiting this note, are we?

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2718/commits/a97a9e24a4c8fb23a5ba007bb7204e5987356b1c

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2718/commits/a97a9e24a4c8fb23a5ba007bb7204e5987356b1c